docs/consistent-hash.md

```diff
@@ -0,0 +1,91 @@
+
+![optimal data redistribution](./images/consistent-hash/data-redistribution-1.svg)
+
+To minimize data movement when scaling occurs, we should be careful when we modify the vnode mapping. Below is an opposite example: modifying vnode mapping like this will result in $\frac{1}{2}$ of the data being moved.
+
+![worst data redistribution](./images/consistent-hash/data-redistribution-2.svg)
+
+### Streaming
+
+We know that a fragment has several actors as its different parallelisms, and that upstream actors will send data to downstream actors via [dispatcher](./streaming-overview.md#actors). The figure below illustrates how actors distribute data based on consistent hash by example. 
+
+![actor data distribution](./images/consistent-hash/actor-data.svg)
+
+In the figure, we can see that one upstream actor dispatches data to three downstream actors. The downstream actors are scheduled on the parallel units mentioned in previous example respectively. 
+
+Based on our consistent hash design, the dispatcher is informed of the latest vnode mapping by meta node. It then decides how to send data by following steps:
+1. Compute vnode of the data via the hash function $H$. Let the vnode be $v_k$.
+2. Look up vnode mapping and find out parallel unit $p_n$ that vnode $v_k$ maps to. 
+3. Send data to the downstream actor that is scheduled on parallel unit $p_n$ (remember that one actor will be scheduled on exactly one parallel unit).
+
+In this way, all actors' data (i.e. actors' states) will be distributed according to the vnode mapping constructed by meta. 
+
+When scaling occurs, actors will be re-scheduled accordingly. By modifying the vnode mapping in meta and make streaming act on the new vnode mapping, we could minimize data movement from following aspects:
+
+- The data of existing actors will not be displaced too much. 
+- The block cache of a compute node will not be invalidated too much.
+
+### Batch
+
+When we perform parallel batch read, we should partition the data for each parallelism in a certain way. Now that we have vnodes corresponding to disjoint sets of data, this naturally forms a data partition pattern. That is to say, one vnode could be viewed as a minimal data partition unit, and we could aggregate several vnodes together to get a larger data partition.
+
+In vnode mapping, one parallel unit will correspond to several vnodes. Therefore, we could use parallel unit as the partition bound. Data falling to one parallel unit in vnode mapping will be one partition. This is better than range partition in that the partition is more stable when the primary key of a materialized view distributes non-randomly.
```

Comment:
> Therefore, we could use parallel unit as the partition bound.

What is use parallel unit as partition bound?

Reply:
The vnodes that mapped to one parallel unit will form a partition group. Lemme give a clearer description about this in the doc.